### ISSUE TITLE
delegation between non-abstract constructors

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-compiler#2139]
@tombentley This kind of thing is allowed:

```ceylon
class Hello {
    String who;
    shared new For(String who) {
        this.who = who;
    }
    shared new World() 
            extends For("world") {}
    print("Hello " + who);
}
```

Currently this results in:

```
Cannot find symbol
  symbol:   variable who
  location: class Hello
```

And variations which don't cause this error call the `print()` line twice.

**Comment by tombentley** [comment]
I think I know the answer to this, given the following:

```ceylon
class Hello {
    String who;
    shared new For(String who) {
        this.who = who;
    }
    shared new World() 
            extends For("world") {
        print("Greetings " + who);
    }
    print("Hello " + who);
}
```

am I right that we would expect this to print "Hello *xxx*" and then "Greetings *xxx*"?

I looked in the spec for some detailed documentation of this execution order stuff in Chapter 8 (specifically 8.2.5), but it doesn't seem to cover constructor delegation within the same class, or the sequence of execution of statements before and after the constructor.

**Comment by gavinking** [comment]
@tombentley If I execute `Hello.World()`, it should print:

```
Greetings xxx
Hello xxx
```

**Comment by tombentley** [comment]
So how is `For` supposed to know it's being called by `World` when it could be being called by anyone else? AFAICS I'm going to have to pass an extra hidden argument to `For` to tell it whether or not to execute each of the possible "footers" (depending on what other constructors are present).

**Comment by gavinking** [comment]
So what are you saying: you want to simply disallow delegation to non-`abstract` constructors?

**Comment by tombentley** [comment]
I don't *want* to, but implementing them correctly does seem to require some kind of caller-awareness. That could surely be implemented just by passing an `int` or something. Using overloading we could avoid exposing that `int`-taking constructor to Java clients.

But I'd be interested to see cases where delegating to a non-`abstract` constructor offered something that couldn't be achieved in other ways.

**Comment by gavinking** [comment]
I'm OK with adding that restriction for now. Let me do it.

**Comment by gavinking** [comment]
Sucks, because I now can't write stuff like this:

```ceylon
class Point {
    shared Float x; shared Float y;
    shared new (Float x, Float y) {
        this.x = x; this.y = y;
    }
    shared new Diagonal(Float d) 
        extends Point(d/sqrt(2), d/sqrt(2)) {}
}
```

**Comment by gavinking** [comment]
This restriction is implemented in https://github.com/ceylon/ceylon-spec/commit/94688b378c84c4263845b53931c1759e5ae5ef98.

I hate it, FTR.

**Comment by gavinking** [comment]
@tombentley couldn't you solve this problem be generating _two _ copies of the non-partial constructor? One would be used for delegation, and one for instantiation.

**Comment by tombentley** [comment]
Yes, that's basically what I suggested (or at least what I meant) above, but the thing is I need the one used for delegation to have a parameter to control which footer statements should be exexcuted. 

```ceylon
class Example{
    shared new ConcreteSelf() {}
    print("a");
    shared new X1()extends ConcreteSelf() {}
    print("b");
    shared new X2() extends ConcreteSelf() {}
    print("c");
}
```

would translate to something like:

```
class Example {
    ConcreteSelf(int x) {
        if ((x&1) != 0) { print("a"); }
        if ((x&2) != 0) { print("a"); }
        if ((x&4) != 0) { print("a"); }
    }
    public ConcreteSelf() {
        this(1|2|4);
    }
    public X1() {
        this(1);
        print("b");
        print("c");
    }
    public X2() {
        this(1|2);
        print("c");
    }
}
```

So depending on what constructors there are I can selectively execute of not depending on the caller. Since the set of calling constructors is known at compile time a bit set is sufficient to encode this.

**Comment by quintesse** [comment]
Dunno, seems to me that if you're going to go this way why not go all the way and do something like:

```ceylon
class Example{
    shared new ConcreteSelf() { print("a"); }
    print("b");
    shared new X1()extends ConcreteSelf() { print("c"); }
    print("d");
    shared new X2() extends ConcreteSelf() { print("e"); }
    print("f");
}
```

```java
class Example {
    private ConcreteSelf(int x) {
        if ((x&1) != 0) { print("a"); }
        if ((x&2) != 0) { print("b"); }
        if ((x&4) != 0) { print("c"); }
        if ((x&8) != 0) { print("d"); }
        if ((x&16) != 0) { print("e"); }
        if ((x&32) != 0) { print("f"); }
    }
    public ConcreteSelf() {
        this(1|2|8|32);
    }
    public X1() {
        this(2|4|8|32);
    }
    public X2() {
        this(2|8|16|32);
    }
}
```

**Comment by gavinking** [comment]
@tombentley  Nononono, that's just nasty, and AFAIK, unnecessary.

Look, the following already works correctly, without any weird bit-matching:

```ceylon
class Example {
    print("1");
    abstract new ConcreteSelf() {}
    print("2");
    shared new X1() extends ConcreteSelf() {
        print("2.5");
    }
    print("3");
    shared new X2() extends ConcreteSelf() {
        print("3.5");
    }
    print("4");
}
```

So, why not treat this:

```ceylon
class Example {
    print("1");
    shared new ConcreteSelf() {}
    print("2");
    shared new X1() extends ConcreteSelf() {
        print("2.5");
    }
    print("3");
    shared new X2() extends ConcreteSelf() {
        print("3.5");
    }
    print("4");
}
```

As the same as this:

```ceylon
class Example {
    print("1");
    abstract new ConcreteSelf_() {}
    shared new ConcreteSelf() extends ConcreteSelf_() {}
    print("2");
    shared new X1() extends ConcreteSelf() {
        print("2.5");
    }
    print("3");
    shared new X2() extends ConcreteSelf() {
        print("3.5");
    }
    print("4");
}
```

**Comment by gavinking** [comment]
No no no no no this bit matching rubbish is crazy I don't know where that's coming from but it's nuts. Please just drop it.

**Comment by chochos** [comment]
I was writing most of the same stuff @tombentley already wrote here, in the js issue. Extending this last example:

```ceylon
class Example {
    print("1");
    shared new ConcreteSelf() {
        print("1.5");
    }
    print("2");
    shared new X1() extends ConcreteSelf() {
        print("2.5");
    }
    print("3");
    shared new X2() extends ConcreteSelf() {
        print("3.5");
    }
    print("4");
}
```

Then `ConcreteSelf` should print `1, 1.5, 2, 3, 4`; `X1` should print `1, 1.5, 2, 2.5, 3, 4` and `X2` should print `1, 1.5, 2, 3, 3.5, 4`. But how can `ConcreteSelf` know not to print 2, 3, 4?

If we transform `ConcreteSelf` into two constructors I guess it should end up like

```ceylon
abstract new ConcreteSelf_(){
  print("1.5");
}
shared new ConcreteSelf() extends ConcreteSelf_() {}
```

is that right?

**Comment by chochos** [comment]
Ok I just tested that transformation by hand and it works.

**Comment by luolong** [comment]
So let me try to get this right:

Given this class declaration:
```ceylon
class Example {
    print("1");
    shared new ConcreteSelf() {
        print("1.5");
    }
    print("2");
    shared new X1() extends ConcreteSelf() {
        print("2.5");
    }
    print("3");
    shared new X2() extends ConcreteSelf() {
        print("3.5");
    }
    print("4");
}
```

Should be equivalent to writing something like this:
```ceylon
class Example {
    new ConcreteSelf_() {
        print("1");
        print("1.5");
    }
    shared new ConcreteSelf() extends ConcreteSelf_() {
        print("2");
        print("3");
        print("4");
    }
    new X1_() extends ConcreteSelf_() {
        print("2");
        print("2.5");
    }
    shared new X1() extends X1_() {
        print("3");
        print("4");
    }
    new X2_() extends ConcreteSelf_() {
        print("3");
        print("3.5");
    }
    shared new X2() extends X2_() {
        print("4");
    }
}
```

**Comment by gavinking** [comment]
@luolong yes, I think that's right, unless I missed something

**Comment by luolong** [comment]
Now just to throw in another wrench, what happens I I add another class:
```ceylon
class Example2 extends Example {
    print("5");
    shared new Shiny() extends Example.X2() {
        print("5.5");
    }
    print("6");
}
```

**Comment by luolong** [comment]
Actually, now that I think about it, nothing special there... This should _just work_...

**Comment by tombentley** [comment]
I *think* I did this, right?
